src/Servers/Kestrel/Core/src/Internal/Http/Http1Connection.cs

```diff
@@ -191,32 +191,77 @@ public bool TakeStartLine(ReadOnlySequence<byte> buffer, out SequencePosition co
             return result;
         }
 
-        public bool TakeMessageHeaders(ReadOnlySequence<byte> buffer, out SequencePosition consumed, out SequencePosition examined)
+        public bool TakeMessageHeaders(in ReadOnlySequence<byte> buffer, out SequencePosition consumed, out SequencePosition examined)
         {
             // Make sure the buffer is limited
-            bool overLength = false;
-            if (buffer.Length >= _remainingRequestHeadersBytesAllowed)
+            if (buffer.Length > _remainingRequestHeadersBytesAllowed)
             {
-                buffer = buffer.Slice(buffer.Start, _remainingRequestHeadersBytesAllowed);
-
-                // If we sliced it means the current buffer bigger than what we're
-                // allowed to look at
-                overLength = true;
+                return TrimAndTakeMessageHeaders(buffer, out consumed, out examined);
             }
 
-            var result = _parser.ParseHeaders(new Http1ParsingHandler(this), buffer, out consumed, out examined, out var consumedBytes);
-            _remainingRequestHeadersBytesAllowed -= consumedBytes;
-
-            if (!result && overLength)
+            var reader = new SequenceReader<byte>(buffer);
+            var result = false;
+            try
             {
-                BadHttpRequestException.Throw(RequestRejectionReason.HeadersExceedMaxTotalSize);
+                result = _parser.ParseHeaders(new Http1ParsingHandler(this), ref reader);
+
+                if (result)
+                {
+                    TimeoutControl.CancelTimeout();
+                }
+
+                return result;
             }
-            if (result)
+            finally
             {
-                TimeoutControl.CancelTimeout();
+                consumed = reader.Position;
+                _remainingRequestHeadersBytesAllowed -= (int)reader.Consumed;
+
+                if (result)
+                {
+                    examined = consumed;
+                }
+                else
+                {
+                    examined = buffer.End;
+                }
             }
 
-            return result;
+            // TODO: In C#8 we can shadow the param and var names and don't have to change them
+            bool TrimAndTakeMessageHeaders(in ReadOnlySequence<byte> buf, out SequencePosition cons, out SequencePosition exam)
```

Comment:
We should have C# 8 now.

Reply:
This line disagrees https://github.com/aspnet/AspNetCore/blob/4aad7548c1b7cf16014e0d543b155f90ce6a8ab5/eng/targets/CSharp.Common.props#L4

Reply:
But https://github.com/aspnet/AspNetCore/blob/46fe5956066c6d665023e7f6cacd5de8c5918dd2/src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs#L407. LAME. We should have updated it for everyone. OK we should fix this.

Reply:
Added PR https://github.com/aspnet/AspNetCore/pull/8128

Reply:
Using C#8 now